lib/pure/views.nim

```diff
@@ -0,0 +1,148 @@
+## View is a type representing a range of elements in an array. In can be thought as a pointer plus a size.
+## View can be created to an arbitrary memory segment and can additionally keep a single ``ref`` object alive.
+##
+## This module defines views and several helper operations on them. All functions in this module (except for unsafe version of ``initView``) are designed to be memory safe (e.g. they raise exception on out-of-bounds accesses).
+import strutils
+
+const
+  needGcKeep = not (compileOption("gc", "boehm") or compileOption("gc", "none"))
+
+type
+  View*[T] = object
+    data: ptr T
+    length: int
+    when needGcKeep:
+      gcKeep: RootRef
+
+  ByteView* = View[byte]
+
+proc initView*[T; R: ref](data: ptr T, len: int, gcKeep: R): View[T] =
+  ## Creates a new view pointing, starting at ``data`` with length ``len``.
+  ##
+  ## ``gcKeep`` should be a pointer to the underlying GC objects that contains these items.
+  when needGcKeep:
+    return View[T](data: data, length: len, gcKeep: cast[RootRef](gcKeep))
+  else:
+    return View[T](data: data, length: len)
+
+proc initView*[T](data: ptr T, len: int): View[T] =
+  ## Creates a new view, starting at ``data`` with length ``len``.
+  ##
+  ## You need to ensure that data pointed by this view is not garbage collected while using the view.
+  when needGcKeep:
+    return View[T](data: data, length: len, gcKeep: nil)
+  else:
+    return View[T](data: data, length: len)
+
+proc initEmptyView*[T](typ: typedesc[T]): View[T] =
+  ## Creates a view of length zero.
+  return initView[T](nil, 0)
+
+proc initView[T](s: ref seq[T]): View[T] =
+  ## Returns a view into a sequence.
+  if s[].len == 0:
+    initEmptyView(T)
+  else:
+    initView(addr s[0], s[].len, gcKeep=s)
+
+proc newView*[T](s: seq[T]): View[T] =
+  ## Copies a sequence and returns a new view pointing into the copy.
+  let copied = new(seq[T])
+  copied[] = s
+  result = initView(copied)
+
+proc newView*[T](typ: typedesc[T], len: int): View[T] =
+  ## Create uninitialized view of length ``len``.
+  let s = new(seq[T])
+  s[] = newSeq[T](len)
+  result = initView(s)
+
+proc isNil*(v: View): bool =
+  return v.len == 0
```

Comment:
This is rather misleading.

Reply:
Also, how can you use 'len' here when it's defined below this function?

Reply:
View is generic and len has multiple overloads, so it's `mixin` by default.

Reply:
I would say, a View is never Nil, just empty and therefore tthis `isNil` should be removed.